src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/ChainPal.cs

```diff
@@ -56,106 +55,62 @@ public static bool ReleaseSafeX509ChainHandle(IntPtr handle)
-                    leaf,
-                    candidates,
-                    systemTrusted,
-                    applicationPolicy,
-                    certificatePolicy,
-                    revocationMode,
-                    revocationFlag,
-                    verificationTime,
-                    ref remainingDownloadTime);
+            OpenSslX509ChainProcessor chainPal = OpenSslX509ChainProcessor.InitiateChain(
+                ((OpenSslX509CertificateReader)cert).SafeHandle,
+                verificationTime,
+                remainingDownloadTime);
 
-#if DEBUG
-                if (chain.ChainElements.Length > 0)
-                {
-                    X509Certificate2 reportedLeaf = chain.ChainElements[0].Certificate;
-                    Debug.Assert(reportedLeaf != null, "reportedLeaf != null");
-                    Debug.Assert(reportedLeaf.Equals(leaf), "reportedLeaf.Equals(leaf)");
-                    Debug.Assert(!ReferenceEquals(reportedLeaf, leaf), "!ReferenceEquals(reportedLeaf, leaf)");
-                }
-#endif
-
-                if (chain.ChainStatus.Length == 0 && downloaded.Count > 0)
-                {
-                    SaveIntermediateCertificates(chain.ChainElements, downloaded);
-                }
+            Interop.Crypto.X509VerifyStatusCode status = chainPal.FindFirstChain(extraStore);
 
-                // Everything we put into the chain has been cloned, dispose all the originals.
-                systemTrusted.DisposeAll();
-                downloaded.DisposeAll();
+            if (!OpenSslX509ChainProcessor.IsCompleteChain(status))
+            {
+                List<X509Certificate2> tmp = null;
+                status = chainPal.FindChainViaAia(ref tmp);
 
-                if (extraStore == null || extraStore.Count == 0)
+                if (tmp != null)
                 {
-                    // There were no extraStore certs, so everything can be disposed.
-                    foreach (X509Certificate2 candidate in candidates)
-                    {
-                        candidate.Dispose();
-                    }
-                }
-                else
-                {
-                    // Candidate certs which came from extraStore should NOT be disposed, since they came
-                    // from outside.
-                    var extraStoreByReference = new HashSet<X509Certificate2>(
-                        ReferenceEqualityComparer<X509Certificate2>.Instance);
-
-                    foreach (X509Certificate2 extraCert in extraStore)
+                    if (status == Interop.Crypto.X509VerifyStatusCode.X509_V_OK)
                     {
-                        extraStoreByReference.Add(extraCert);
+                        SaveIntermediateCertificates(tmp);
                     }
 
-                    foreach (X509Certificate2 candidate in candidates)
+                    foreach (X509Certificate2 downloaded in tmp)
```

Comment:
should this be in try .. finally?

Reply:
It could be, but once the chain building starts there aren't very many things that would throw (most of the places that can throw will only throw when malloc starts failing).  Once something catastrophic is happening I think I'm okay with letting one or two cert handles move to the finalizer queue.  But I'm open to alternate opinions.